### ISSUE TITLE
Feature: Repeat Question Groups

### ISSUE THREAD

**Issue by ichinaski** [akvo/akvo-flow-mobile#257]
# Overview

The aim is to introduce repeated question groups. Often, a number of questions need to be asked multiple times, for example if you want to ask a number of questions to each member of a household. This functionality lets the user specify that a question group will be repeated a number of times.

# Docs & requirements

* [Functional design](https://github.com/akvo/akvo-product-design/blob/master/FLOW/Features/35-RepeatQuestionGroups/FunctionalDesign.md)
* [Technical design](https://github.com/akvo/akvo-product-design/blob/master/FLOW/Features/35-RepeatQuestionGroups/TechnicalDesign.md)
* akvo-flow issue: https://github.com/akvo/akvo-flow/issues/269

# Progress

* [x] Duplicate question views, creating a unique identifier for each iteration.
* [x] Save responses into the DB
* [x] Load responses from the DB
* [ ] Handle dependencies:
    * Cross-tab dependencies
    * Repeatable question dependencies
* [ ] Export JSON data
* [ ] Import JSON data (monitoring surveys)
* [ ] UI overhaul: Style, UX actions, etc

**Comment by ichinaski** [comment]
### Test plan

**1. Basics:**
* Create a repeatable question group. No dependencies.
* Repeat and delete iterations. Indices should be accordingly updated.
* Check the `iteration` IDs in the datastore. They should reflect the order in which they were created, starting at 0, and without taking the deleted iterations into account (i.e. if there are any deleted iterations, the IDs won't match to the visual numbering in the app).

**2. Dependecies:**
* Create a form with:
    * A regular group *Group 1*, containing an option question.
    * A repeatable group *Group 2*, wherein one (or more) question depends on the first group's option question.
    * A repeatable group *Group 3*, containing an option question, wherein one (or more) question depends on this option question (same group).
* For each repetition in *Group 2*, the dependent questions should be enabled/disabled according to the parent question's response (in *Group 1*). Note that all iterations must be updated if the response is changed.
* Dependencies in *Group 3* are localized to any given iteration. Trigger the parent question's response, and only the dependent question **in this group* should reflect any change.
* Ensure hidden (not visible) questions don't make it to the dashboard.

**Comment by iperdomo** [comment]
Hi, we found an issue while testing this functionality in the app. Here you have the steps to reproduce:

* In the dashboard create one question group and mark it as _repeatable_
* Add 2 _required_ questions to that group
* Assign it to the device and try to fill in the form
* Start filling the form, and make a couple of iterations, but leave empty one of the required questions in one of the iterations
* Move to the Submit screen
* In the _Submit_ screen, you get the list of unanswered but required questions,
* Click the _Pen_ icon that takes you back to the question that is required but unanswered, __the Question Group  tab gets rendered as an empty page__

cc/ @lauraroverts

**Comment by ichinaski** [comment]
@iperdomo good catch. I found the problem and provided a fix. I have also updated the app at akvoflow-80, in case you need it for the training.

**Comment by rumca** [comment]
The above test plans pass as described - tried a few other combinations of dependencies and everything seemed to play well too :+1:

**Comment by rumca** [comment]
@ichinaski I checked a bunch of other things with this functionality whilst working on the 1.9.2 testing. I noticed a few peculiarities (probably both the same underlying cause):

* When a question has the datapoint name flag set and the question group is repeatable, the name is not set as expected. It instead remains as "Unnamed datapoint". Updating the question group to remove the repeatable property results in subsequent responses being populated as expected.
* When there's a geolocation question and the group is repeatable, the responses are not appearing on the map in the dashboard as expected. I assume having a geolocation question in a RQG would be highly unusual - but might be worth checking out just in case.

Let me know if I should open a new issue for the above.
